File: Aran/Panda/EnRoute/Modules/Navaids_DataBase.vb
Option Strict Off
Option Explicit On

<System.Runtime.InteropServices.ComVisibleAttribute(False)> Module Navaids_DataBase
	<System.Runtime.InteropServices.ComVisibleAttribute(False)> Public Enum eNavaidType
		NONE = -1
		VOR = 0
		DME = 1
		NDB = 2
		LLZ = 3
		TACAN = 4
		FIX = 5
		MKR = 6
	End Enum

	Public Const TACANDME As eNavaidType = CType(128, eNavaidType)
	Public Const TACANVOR As eNavaidType = CType(256, eNavaidType)

	Public NavTypeNames() As String = {"VOR", "DME", "NDB", "LOC", "TACAN", "Radar FIX", "MKR"}

	<System.Runtime.InteropServices.ComVisibleAttribute(False)> Structure VORData
		Dim Range As Double
		Dim FA_Range As Double
		Dim InitWidth As Double
		Dim SplayAngle As Double
		Dim TrackingTolerance As Double
		Dim IntersectingTolerance As Double
		Dim ConeAngle As Double
		Dim TrackAccuracy As Double
		Dim LateralDeviationCoef As Double
		Dim EnRouteTrackingToler As Double
		Dim EnRouteTracking2Toler As Double
		Dim EnRouteInterToler As Double
		Dim EnRoutePrimAreaWith As Double
		Dim EnRouteSecAreaWith As Double
		Dim OnNAVRadius As Double
	End Structure

	<System.Runtime.InteropServices.ComVisibleAttribute(False)> Structure NDBData
		Dim Range As Double
		Dim FA_Range As Double
		Dim InitWidth As Double
		Dim SplayAngle As Double
		Dim TrackingTolerance As Double
		Dim IntersectingTolerance As Double
		Dim ConeAngle As Double
		Dim TrackAccuracy As Double
		Dim Entry2ConeAccuracy As Double
		Dim LateralDeviationCoef As Double
		Dim EnRouteTrackingToler As Double
		Dim EnRouteTracking2Toler As Double
		Dim EnRouteInterToler As Double
		Dim EnRoutePrimAreaWith As Double
		Dim EnRouteSecAreaWith As Double
		Dim OnNAVRadius As Double
	End Structure

	<System.Runtime.InteropServices.ComVisibleAttribute(False)> Structure DMEData
		Dim Range As Double
		Dim MinimalError As Double
		Dim ErrorScalingUp As Double
		Dim SlantAngle As Double
		Dim TP_div As Double
	End Structure

	<System.Runtime.InteropServices.ComVisibleAttribute(False)> Structure LLZData
		Dim Range As Double
		Dim TrackingTolerance As Double
		Dim IntersectingTolerance As Double
	End Structure

	<System.Runtime.InteropServices.ComVisibleAttribute(False)> Structure TACANData
		Dim DMECon As DMEData
		Dim VORCon As VORData
	End Structure

	Public NavaidTypesCount As Integer
	Public NavaidTypes As ESRI.ArcGIS.Geodatabase.ITable
	Public Navaids() As ESRI.ArcGIS.Geodatabase.ITable

	Public VOR As VORData
	Public NDB As NDBData
	Public DME As DMEData
	Public LLZ As LLZData
	Public TACAN As TACANData

	Function InitModule() As Boolean
		Dim I As Integer
		Dim N As Integer
		Dim J As Integer
		Dim iBaseName As Integer
		Dim iNavTypeName As Integer
		Dim iParam_Name As Integer
		Dim iValue As Integer
		Dim iUnit As Integer
		Dim iMultiplier As Integer

		Dim Value As Double
		Dim Multiplier As Double
		Dim ParName As String

		Dim pTable As ESRI.ArcGIS.Geodatabase.ITable = Nothing
		Dim pRow As ESRI.ArcGIS.Geodatabase.IRow

		On Error GoTo EH

		If Not OpenTableFromFile(NavaidTypes, ConstDir + "\Navaids\", "Navaids") Then Return False

		iBaseName = NavaidTypes.FindField("BaseName")
		iNavTypeName = NavaidTypes.FindField("Name")

		NavaidTypesCount = NavaidTypes.RowCount(Nothing)
		ReDim Navaids(NavaidTypesCount - 1)

		For J = 0 To NavaidTypesCount - 1
			If Not OpenTableFromFile(pTable, ConstDir + "\Navaids\", NavaidTypes.GetRow(J).Value(iBaseName)) Then Return False

			Navaids(J) = pTable

			Select Case NavaidTypes.GetRow(J).Value(iNavTypeName)
				'============================ VOR ====================================
				Case "VOR"
					iParam_Name = -1
					iValue = -1
					iUnit = -1
					iMultiplier = -1

					For I = 0 To Navaids(J).Fields.FieldCount - 1
						If Navaids(J).Fields.Field(I).Name = "PARAM_NAME" Then iParam_Name = I
						If Navaids(J).Fields.Field(I).Name = "VALUE" Then iValue = I
						If Navaids(J).Fields.Field(I).Name = "UNIT" Then iUnit = I
						If Navaids(J).Fields.Field(I).Name = "MULTIPLIER" Then iMultiplier = I
					Next I

					N = Navaids(J).RowCount(Nothing)

					For I = 0 To N - 1
						pRow = Navaids(J).GetRow(I)
						ParName = pRow.Value(iParam_Name)
						Multiplier = pRow.Value(iMultiplier)
						Value = pRow.Value(iValue) * Multiplier
						If (pRow.Value(iUnit) = "rad") Or (pRow.Value(iUnit) = "°") Then Value = System.Math.Round(RadToDeg(Value), 2)

						If ParName = "Range" Then VOR.Range = Value
						If ParName = "FA Range" Then VOR.FA_Range = Value
						If ParName = "Initial width" Then VOR.InitWidth = Value
						If ParName = "Splay angle" Then VOR.SplayAngle = Value
						If ParName = "Tracking tolerance" Then VOR.TrackingTolerance = Value
						If ParName = "Intersecting tolerance" Then VOR.IntersectingTolerance = Value
						If ParName = "Cone angle" Then VOR.ConeAngle = Value
						If ParName = "Track accuracy" Then VOR.TrackAccuracy = Value
						If ParName = "Lateral deviation coef." Then VOR.LateralDeviationCoef = Value
						If ParName = "EnRoute Tracking toler" Then VOR.EnRouteTrackingToler = Value
						If ParName = "EnRoute Tracking2 toler" Then VOR.EnRouteTracking2Toler = Value
						If ParName = "EnRoute Inter toler" Then VOR.EnRouteInterToler = Value
						If ParName = "EnRoute PrimArea With" Then VOR.EnRoutePrimAreaWith = Value
						If ParName = "EnRoute SecArea With" Then VOR.EnRouteSecAreaWith = Value
					Next I
					'============================ NDB ====================================
				Case "NDB"
					iParam_Name = -1
					iValue = -1
					iUnit = -1
					iMultiplier = -1

					For I = 0 To Navaids(J).Fields.FieldCount - 1
						If Navaids(J).Fields.Field(I).Name = "PARAM_NAME" Then iParam_Name = I
						If Navaids(J).Fields.Field(I).Name = "VALUE" Then iValue = I
						If Navaids(J).Fields.Field(I).Name = "UNIT" Then iUnit = I
						If Navaids(J).Fields.Field(I).Name = "MULTIPLIER" Then iMultiplier = I
					Next I

					N = Navaids(J).RowCount(Nothing)

					For I = 0 To N - 1
						pRow = Navaids(J).GetRow(I)
						ParName = pRow.Value(iParam_Name)
						Multiplier = pRow.Value(iMultiplier)
						Value = pRow.Value(iValue) * Multiplier
						If (pRow.Value(iUnit) = "rad") Or (pRow.Value(iUnit) = "°") Then Value = System.Math.Round(RadToDeg(Value), 2)

						If ParName = "Range" Then NDB.Range = Value
						If ParName = "FA Range" Then NDB.FA_Range = Value
						If ParName = "Initial width" Then NDB.InitWidth = Value
						If ParName = "Splay angle" Then NDB.SplayAngle = Value
						If ParName = "Tracking tolerance" Then NDB.TrackingTolerance = Value
						If ParName = "Intersecting tolerance" Then NDB.IntersectingTolerance = Value
						If ParName = "Cone angle" Then NDB.ConeAngle = Value
						If ParName = "Track accuracy" Then NDB.TrackAccuracy = Value
						If ParName = "Entry into the cone accuracy" Then NDB.Entry2ConeAccuracy = Value
						If ParName = "Lateral deviation coef." Then NDB.LateralDeviationCoef = Value
						If ParName = "EnRoute Tracking toler" Then NDB.EnRouteTrackingToler = Value
						If ParName = "EnRoute Tracking2 toler" Then NDB.EnRouteTracking2Toler = Value
						If ParName = "EnRoute Inter toler" Then NDB.EnRouteInterToler = Value
						If ParName = "EnRoute PrimArea With" Then NDB.EnRoutePrimAreaWith = Value
						If ParName = "EnRoute SecArea With" Then NDB.EnRouteSecAreaWith = Value
					Next I
					'============================ DME ====================================
				Case "DME"
					iParam_Name = -1
					iValue = -1
					iUnit = -1
					iMultiplier = -1

					For I = 0 To Navaids(J).Fields.FieldCount - 1
						If Navaids(J).Fields.Field(I).Name = "PARAM_NAME" Then iParam_Name = I
						If Navaids(J).Fields.Field(I).Name = "VALUE" Then iValue = I
						If Navaids(J).Fields.Field(I).Name = "UNIT" Then iUnit = I
						If Navaids(J).Fields.Field(I).Name = "MULTIPLIER" Then iMultiplier = I
					Next I

					N = Navaids(J).RowCount(Nothing)

					For I = 0 To N - 1
						pRow = Navaids(J).GetRow(I)
						ParName = pRow.Value(iParam_Name)
						Multiplier = pRow.Value(iMultiplier)
						Value = pRow.Value(iValue) * Multiplier
						If (pRow.Value(iUnit) = "rad") Or (pRow.Value(iUnit) = "°") Then Value = Math.Round(RadToDeg(Value), 2)

						If ParName = "Range" Then DME.Range = Value
						If ParName = "Minimal Error" Then DME.MinimalError = Value
						If ParName = "Error Scaling Up" Then DME.ErrorScalingUp = Value
						If ParName = "Slant Angle" Then DME.SlantAngle = Value
						If ParName = "TP_div" Then DME.TP_div = Value
					Next I
					'============================ LLZ ====================================
				Case "LLZ"
					iParam_Name = -1
					iValue = -1
					iUnit = -1
					iMultiplier = -1

					For I = 0 To Navaids(J).Fields.FieldCount - 1
						If Navaids(J).Fields.Field(I).Name = "PARAM_NAME" Then iParam_Name = I
						If Navaids(J).Fields.Field(I).Name = "VALUE" Then iValue = I
						If Navaids(J).Fields.Field(I).Name = "UNIT" Then iUnit = I
						If Navaids(J).Fields.Field(I).Name = "MULTIPLIER" Then iMultiplier = I
					Next I

					N = Navaids(J).RowCount(Nothing)

					For I = 0 To N - 1
						pRow = Navaids(J).GetRow(I)
						ParName = pRow.Value(iParam_Name)
						Multiplier = pRow.Value(iMultiplier)
						Value = pRow.Value(iValue) * Multiplier
						If (pRow.Value(iUnit) = "rad") Or (pRow.Value(iUnit) = "°") Then Value = Math.Round(RadToDeg(Value), 2)

						If ParName = "Range" Then LLZ.Range = Value
						If ParName = "Tracking tolerance" Then LLZ.TrackingTolerance = Value
						If ParName = "Intersecting tolerance" Then LLZ.IntersectingTolerance = Value
					Next I
					'============================ FIN ====================================
			End Select
		Next J

		TACAN.DMECon = DME
		TACAN.VORCon = VOR

		VOR.OnNAVRadius = System.Math.Sin(DegToRad(VOR.TrackAccuracy + TrackAccuracy.Value)) / System.Math.Sin(DegToRad(90.0 - TrackAccuracy.Value)) * OverHeadToler.Value
		NDB.OnNAVRadius = System.Math.Sin(DegToRad(NDB.Entry2ConeAccuracy + TrackAccuracy.Value)) / System.Math.Sin(DegToRad(90.0 - TrackAccuracy.Value)) * OverHeadToler.Value

		Return True
EH:
		MessageBox.Show(Err.Number + "  " + Err.Description)
		Return False
	End Function

End Module

File: Aran/Panda/Departure/trunk/Modules/Navaids_DataBase.cs
using System;
using ESRI.ArcGIS.Geodatabase;
using ESRI.ArcGIS.Geometry;

namespace Aran.PANDA.Departure
{
	[System.Runtime.InteropServices.ComVisible(false)]
	public enum eNavaidType
	{
		NONE = -1,
		VOR = 0,
		DME = 1,
		NDB = 2,
		LLZ = 3,
		TACAN = 4,
		//StartPoint = 5
	}

	[System.Runtime.InteropServices.ComVisible(false)]
	public static class Navaids_DataBase
	{
		public static ITable NavaidTypes;

		public static VORData VOR = new VORData();
		public static NDBData NDB = new NDBData();
		public static DMEData DME = new DMEData();
		public static LLZData LLZ = new LLZData();

		public static string[] NavTypeNames = new string[] { "VOR", "DME", "NDB", "LOC", "TACAN", "Radar FIX" };

		public static string GetNavTypeName(eNavaidType navType)
		{
			if (navType == eNavaidType.NONE)
				return "WPT";
			else
				return NavTypeNames[(int)navType];
		}

		public static void InitModule()
		{
			int i, j, n;
			int NavaidTypesCount,
				iBaseName, iNavTypeName, iParam_Name, iValue, iUnit, iMultiplier;

			double Value, Multiplier;
			string ParamName;

			ITable pNavaidTable;
			IRow pRow;

			NavaidTypes = FileRoutines.OpenTableFromFile(GlobalVars.ConstDir + @"\Navaids\", "Navaids");

			iBaseName = NavaidTypes.FindField("BaseName");
			iNavTypeName = NavaidTypes.FindField("Name");

			NavaidTypesCount = NavaidTypes.RowCount(null);

			for (j = 0; j < NavaidTypesCount; j++)
			{
				pNavaidTable = FileRoutines.OpenTableFromFile(GlobalVars.ConstDir + @"\Navaids\",
					NavaidTypes.GetRow(j).get_Value(iBaseName).ToString());

				string sNavTypeName = NavaidTypes.GetRow(j).get_Value(iNavTypeName).ToString();
				switch (sNavTypeName)
				{
					case "VOR":
						iParam_Name = -1;
						iValue = -1;
						iUnit = -1;
						iMultiplier = -1;

						for (i = 0; i < pNavaidTable.Fields.FieldCount; i++)
						{
							if (pNavaidTable.Fields.get_Field(i).Name == "PARAM_NAME")
								iParam_Name = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "VALUE")
								iValue = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "UNIT")
								iUnit = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "MULTIPLIER")
								iMultiplier = i;

						}

						n = pNavaidTable.RowCount(null);

						for (i = 0; i < n; i++)
						{
							pRow = pNavaidTable.GetRow(i);
							ParamName = pRow.Value[iParam_Name].ToString();
							Multiplier = System.Convert.ToDouble(pRow.Value[iMultiplier]);
							Value = System.Convert.ToDouble(pRow.Value[iValue]) * Multiplier;

							if (pRow.Value[iUnit].ToString() == "rad" || pRow.Value[iUnit].ToString() == "°")
								Value = Math.Round(GlobalVars.RadToDegValue * Value, 1);

							if (ParamName == "Range")
								VOR.Range = Value;

							if (ParamName == "FA Range")
								VOR.FA_Range = Value;

							if (ParamName == "Initial width")
								VOR.InitWidth = Value;

							if (ParamName == "Splay angle")
								VOR.SplayAngle = Value;

							if (ParamName == "Tracking tolerance")
								VOR.TrackingTolerance = Value;

							if (ParamName == "Intersecting tolerance")
								VOR.IntersectingTolerance = Value;

							if (ParamName == "Cone angle")
								VOR.ConeAngle = Value;

							if (ParamName == "Track accuracy")
								VOR.TrackAccuracy = Value;

							if (ParamName == "Lateral deviation coef.")
								VOR.LateralDeviationCoef = Value;

							if (ParamName == "EnRoute Tracking toler")
								VOR.EnRouteTrackingToler = Value;

							if (ParamName == "EnRoute Tracking2 toler")
								VOR.EnRouteTracking2Toler = Value;

							if (ParamName == "EnRoute Inter toler")
								VOR.EnRouteInterToler = Value;

							if (ParamName == "EnRoute PrimArea With")
								VOR.EnRoutePrimAreaWith = Value;

							if (ParamName == "EnRoute SecArea With")
								VOR.EnRouteSecAreaWith = Value;
						}
						// ============================ NDB ====================================
						break;
					case "NDB":
						iParam_Name = -1;
						iValue = -1;
						iUnit = -1;
						iMultiplier = -1;

						for (i = 0; i < pNavaidTable.Fields.FieldCount; i++)
						{
							if (pNavaidTable.Fields.get_Field(i).Name == "PARAM_NAME")
								iParam_Name = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "VALUE")
								iValue = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "UNIT")
								iUnit = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "MULTIPLIER")
								iMultiplier = i;

						}

						n = pNavaidTable.RowCount(null);

						for (i = 0; i < n; i++)
						{
							pRow = pNavaidTable.GetRow(i);
							ParamName = pRow.Value[iParam_Name].ToString();
							Multiplier = System.Convert.ToDouble(pRow.Value[iMultiplier]);
							Value = System.Convert.ToDouble(pRow.Value[iValue]) * Multiplier;

							if (pRow.Value[iUnit].ToString() == "rad" || pRow.Value[iUnit].ToString() == "°")
								Value = Math.Round(GlobalVars.RadToDegValue * Value, 1);

							if (ParamName == "Range")
								NDB.Range = Value;

							if (ParamName == "FA Range")
								NDB.FA_Range = Value;

							if (ParamName == "Initial width")
								NDB.InitWidth = Value;

							if (ParamName == "Splay angle")
								NDB.SplayAngle = Value;

							if (ParamName == "Tracking tolerance")
								NDB.TrackingTolerance = Value;

							if (ParamName == "Intersecting tolerance")
								NDB.IntersectingTolerance = Value;

							if (ParamName == "Cone angle")
								NDB.ConeAngle = Value;

							if (ParamName == "Track accuracy")
								NDB.TrackAccuracy = Value;

							if (ParamName == "Entry into the cone accuracy")
								NDB.Entry2ConeAccuracy = Value;

							if (ParamName == "Lateral deviation coef.")
								NDB.LateralDeviationCoef = Value;

							if (ParamName == "EnRoute Tracking toler")
								NDB.EnRouteTrackingToler = Value;

							if (ParamName == "EnRoute Tracking2 toler")
								NDB.EnRouteTracking2Toler = Value;

							if (ParamName == "EnRoute Inter toler")
								NDB.EnRouteInterToler = Value;

							if (ParamName == "EnRoute PrimArea With")
								NDB.EnRoutePrimAreaWith = Value;

							if (ParamName == "EnRoute SecArea With")
								NDB.EnRouteSecAreaWith = Value;
						}
						// ============================ DME ====================================
						break;
					case "DME":
						iParam_Name = -1;
						iValue = -1;
						iUnit = -1;
						iMultiplier = -1;

						for (i = 0; i < pNavaidTable.Fields.FieldCount; i++)
						{
							if (pNavaidTable.Fields.get_Field(i).Name == "PARAM_NAME")
								iParam_Name = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "VALUE")
								iValue = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "UNIT")
								iUnit = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "MULTIPLIER")
								iMultiplier = i;

						}

						n = pNavaidTable.RowCount(null);

						for (i = 0; i < n; i++)
						{
							pRow = pNavaidTable.GetRow(i);
							ParamName = pRow.Value[iParam_Name].ToString();
							Multiplier = System.Convert.ToDouble(pRow.Value[iMultiplier]);
							Value = System.Convert.ToDouble(pRow.Value[iValue]) * Multiplier;

							if (pRow.Value[iUnit].ToString() == "rad" || pRow.Value[iUnit].ToString() == "°")
								Value = Math.Round(GlobalVars.RadToDegValue * Value, 1);

							if (ParamName == "Range")
								DME.Range = Value;

							if (ParamName == "Minimal Error")
								DME.MinimalError = Value;

							if (ParamName == "Error Scaling Up")
								DME.ErrorScalingUp = Value;

							if (ParamName == "Slant Angle")
								DME.SlantAngle = Value;

							if (ParamName == "TP_div")
								DME.TP_div = Value;
						}
						// ============================ LLZ ====================================
						break;
					case "LLZ":
						iParam_Name = -1;
						iValue = -1;
						iUnit = -1;
						iMultiplier = -1;

						for (i = 0; i < pNavaidTable.Fields.FieldCount; i++)
						{
							if (pNavaidTable.Fields.get_Field(i).Name == "PARAM_NAME")
								iParam_Name = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "VALUE")
								iValue = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "UNIT")
								iUnit = i;

							if (pNavaidTable.Fields.get_Field(i).Name == "MULTIPLIER")
								iMultiplier = i;
						}

						n = pNavaidTable.RowCount(null);

						for (i = 0; i < n; i++)
						{
							pRow = pNavaidTable.GetRow(i);
							ParamName = pRow.Value[iParam_Name].ToString();
							Multiplier = System.Convert.ToDouble(pRow.Value[iMultiplier]);
							Value = System.Convert.ToDouble(pRow.Value[iValue]) * Multiplier;

							if (pRow.Value[iUnit].ToString() == "rad" || pRow.Value[iUnit].ToString() == "°")
								Value = Math.Round(GlobalVars.RadToDegValue * Value, 1);

							if (ParamName == "Range")
								LLZ.Range = Value;

							if (ParamName == "Tracking tolerance")
								LLZ.TrackingTolerance = Value;

							if (ParamName == "Intersecting tolerance")
								LLZ.IntersectingTolerance = Value;
						}
						// ============================ FIN ====================================
						break;
				}
			}
		}

		//public static int NavNameToID(string Name)
		//{
		//    for (int i = 0; i < GlobalVars.NavaidList.Length; i++)
		//        if (GlobalVars.NavaidList[i].Name == Name)
		//            return i;

		//    return -1;
		//}

		public static double VORFIXTolerArea(IPoint ptVor, double Aztin, double AbsH, out IPointCollection TolerArea)
		{
			double vORFIXTolerAreaReturn = 0;
			double R = 0;
			IPoint ptTmp = null;
			IPoint ptCur = null;
			double fTmp = 0;
			double fTmpH = 0;

			fTmpH = AbsH - ptVor.Z;
			R = fTmpH * System.Math.Tan(Functions.DegToRad(VOR.ConeAngle));

			TolerArea = new Polygon();
			ptTmp = Functions.PointAlongPlane(ptVor, Aztin - (90.0 + VOR.TrackAccuracy), VOR.LateralDeviationCoef * fTmpH);
			fTmp = Functions.CircleVectorIntersect(ptVor, R, ptTmp, Aztin - VOR.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);
			fTmp = Functions.CircleVectorIntersect(ptVor, R, ptTmp, 180.0 + Aztin - VOR.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);
			ptTmp = Functions.PointAlongPlane(ptVor, Aztin + 90.0 + VOR.TrackAccuracy, VOR.LateralDeviationCoef * fTmpH);
			fTmp = Functions.CircleVectorIntersect(ptVor, R, ptTmp, 180.0 + Aztin + VOR.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);

			fTmp = Functions.CircleVectorIntersect(ptVor, R, ptTmp, Aztin + VOR.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);

			vORFIXTolerAreaReturn = R;
			return vORFIXTolerAreaReturn;
		}

		public static double NDBFIXTolerArea(IPoint ptNDB, double Aztin, double AbsH, out IPointCollection TolerArea)
		{
			double nDBFIXTolerAreaReturn = 0;
			double R = 0;
			double qN = 0;
			IPoint ptTmp = null;
			IPoint ptCur = null;
			double fTmp = 0;
			double fTmpH = 0;

			fTmpH = AbsH - ptNDB.Z;
			R = fTmpH * System.Math.Tan(Functions.DegToRad(NDB.ConeAngle));

			TolerArea = new Polygon();

			qN = R * System.Math.Sin(Functions.DegToRad(NDB.Entry2ConeAccuracy));
			ptTmp = Functions.PointAlongPlane(ptNDB, Aztin - 90.0, qN + System.Math.Sqrt(R * R - qN * qN) * System.Math.Tan(Functions.DegToRad(NDB.TrackAccuracy)));
			fTmp = Functions.CircleVectorIntersect(ptNDB, R, ptTmp, Aztin - NDB.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);
			fTmp = Functions.CircleVectorIntersect(ptNDB, R, ptTmp, 180.0 + Aztin - NDB.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);
			ptTmp = Functions.PointAlongPlane(ptNDB, Aztin + 90.0, qN + System.Math.Sqrt(R * R - qN * qN) * System.Math.Tan(Functions.DegToRad(NDB.TrackAccuracy)));
			fTmp = Functions.CircleVectorIntersect(ptNDB, R, ptTmp, 180.0 + Aztin + NDB.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);
			fTmp = Functions.CircleVectorIntersect(ptNDB, R, ptTmp, Aztin + NDB.TrackAccuracy, out ptCur);
			TolerArea.AddPoint(ptCur);
			nDBFIXTolerAreaReturn = R;
			return nDBFIXTolerAreaReturn;
		}

		//NavaidToFIXableNavaid
		//public static NavaidType NavaidType2FixableNavaidType(NavaidType val)
		//{
		//	NavaidType result = new NavaidType();

		//	result.pPtGeo = val.pPtGeo;
		//	result.pPtPrj = val.pPtPrj;
		//	//result.pFeature = val.pFeature;

		//	result.NAV_Ident = val.NAV_Ident;
		//	result.Identifier = val.Identifier;
		//	result.Name = val.Name;
		//	result.CallSign = val.CallSign;
		//	result.MagVar = val.MagVar;
		//	result.HorAccuracy = val.HorAccuracy;

		//	result.TypeCode = val.TypeCode;
		//	result.Range = val.Range;
		//	result.index = val.index;

		//	result.PairNavaidIndex = val.PairNavaidIndex;
		//	result.PairNavaidType = val.PairNavaidType;

		//	result.Tag = val.Tag;

		//	//Res.Front;
		//	//Res.Dist;
		//	//Res.CLShift;
		//	//Res.ValCnt;
		//	//Res.ValMin;
		//	//Res.ValMax;

		//	return result;
		//}

		public static NavaidType WPT_FIXToNavaid(WPT_FIXType val)
		{
			NavaidType result = new NavaidType();

			result.pPtGeo = val.pPtGeo;
			result.pPtPrj = val.pPtPrj;

			result.NAV_Ident = val.NAV_Ident;
			result.Identifier = val.Identifier;
			result.Name = val.Name;
			result.CallSign = val.CallSign;
			result.TypeCode = val.TypeCode;

			result.MagVar = val.MagVar;
			result.HorAccuracy = val.HorAccuracy;

			if (val.TypeCode == eNavaidType.VOR)
				result.Range = VOR.Range;
			else if (val.TypeCode == eNavaidType.NDB)
				result.Range = NDB.Range;
			else if (val.TypeCode == eNavaidType.DME)
				result.Range = DME.Range;
			else
				result.Range = LLZ.Range;

			result.index = -1;
			result.PairNavaidIndex = -1;

			//Res.GP_RDH = 0;
			//Res.Course = 0;
			//Res.LLZ_THR = 0;
			//Res.SecWidth = 0;

			result.ValCnt = -2;
			result.Tag = val.Tag;
			return result;
		}
	}
}
